File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/ConnectionService.java
package org.gozaltech.nvdaremotecompanion.android;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.os.PowerManager;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ConnectionService extends Service {

    private static final String TAG = "NVDARemote/Service";
    private static final int NOTIFICATION_ID = 1;
    private static final String CHANNEL_ID = "nvdaremote_connection";

    public class LocalBinder extends Binder {
        public ConnectionService getService() {
            return ConnectionService.this;
        }
    }

    private final IBinder binder = new LocalBinder();
    private final ExecutorService executor = Executors.newCachedThreadPool();

    private TtsManager ttsManager;
    private NvdaAudioManager audioManager;
    private boolean initialized = false;
    private PowerManager.WakeLock wakeLock;

    private final Set<Integer> desiredConnected =
            Collections.synchronizedSet(new HashSet<>());

    private final ConnectivityManager.NetworkCallback networkCallback =
            new ConnectivityManager.NetworkCallback() {
        @Override
        public void onAvailable(Network network) {
            Set<Integer> toReconnect = new HashSet<>(desiredConnected);
            if (toReconnect.isEmpty()) return;
            Log.i(TAG, "Network available — reconnecting " + toReconnect.size() + " desired profile(s)");
            for (int idx : toReconnect) {
                executor.submit(() -> {
                    if (!NativeBridge.nativeIsConnected(idx)) {
                        boolean ok = NativeBridge.nativeConnect(idx);
                        if (!ok) desiredConnected.remove(idx);
                    }
                });
            }
        }

        @Override
        public void onLost(Network network) {
            Set<Integer> toDisconnect = new HashSet<>(desiredConnected);
            if (toDisconnect.isEmpty()) return;
            Log.i(TAG, "Network lost — disconnecting " + toDisconnect.size() + " profile(s)");
            for (int idx : toDisconnect) {
                executor.submit(() -> {
                    if (NativeBridge.nativeIsConnected(idx))
                        NativeBridge.nativeDisconnect(idx);
                });
            }
        }
    };

    private final NativeBridge.ConnectionStateListener connectionStateListener =
            this::updateNotification;

    @Override
    public void onCreate() {
        super.onCreate();
        createNotificationChannel();

        String enginePkg = AppPrefs.getTtsEngine(this);
        ttsManager = new TtsManager(getApplicationContext(), enginePkg);
        ttsManager.setUseAccessibilityStream(AppPrefs.getAccessibilityStream(this));
        ttsManager.setPitch(AppPrefs.getTtsPitch(this));
        ttsManager.setRate(AppPrefs.getTtsRate(this));
        ttsManager.setVolume(AppPrefs.getTtsVolume(this));
        audioManager = new NvdaAudioManager(getApplicationContext());

        ttsManager.setScreenReaderMode(AppPrefs.getScreenReaderMode(this));
        PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "NVDARemote:connection");

        NativeBridge.setAppContext(this);
        NativeBridge.nativeInit(ttsManager, audioManager, getFilesDir().getAbsolutePath());
        initialized = true;
        Log.i(TAG, "ConnectionService created, native initialized");

        startForeground();
        autoConnect();
        NativeBridge.addConnectionStateListener(connectionStateListener);

        NetworkRequest request = new NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .build();
        getSystemService(ConnectivityManager.class).registerNetworkCallback(request, networkCallback);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onDestroy() {
        NativeBridge.removeConnectionStateListener(connectionStateListener);
        getSystemService(ConnectivityManager.class).unregisterNetworkCallback(networkCallback);
        executor.shutdownNow();
        if (initialized) {
            int count = NativeBridge.nativeGetProfileCount();
            for (int i = 0; i < count; i++) {
                if (NativeBridge.nativeIsConnected(i)) NativeBridge.nativeDisconnect(i);
            }
            NativeBridge.nativeShutdown();
            initialized = false;
        }
        if (wakeLock != null && wakeLock.isHeld()) wakeLock.release();
        ttsManager.shutdown();
        Log.i(TAG, "ConnectionService destroyed");
        super.onDestroy();
    }

    public interface ConnectResultCallback {
        void onResult(boolean success);
    }

    public void connect(int profileIndex, ConnectResultCallback callback) {
        desiredConnected.add(profileIndex);
        executor.submit(() -> {
            boolean ok = NativeBridge.nativeConnect(profileIndex);
            Log.i(TAG, "connect(" + profileIndex + ") = " + ok);
            if (!ok) desiredConnected.remove(profileIndex);
            if (callback != null) callback.onResult(ok);
        });
    }

    public void disconnect(int profileIndex) {
        desiredConnected.remove(profileIndex);
        executor.submit(() -> NativeBridge.nativeDisconnect(profileIndex));
    }

    public void setActiveProfile(int profileIndex) { NativeBridge.nativeSetActiveProfile(profileIndex); }
    public void toggleForwarding()                 { NativeBridge.nativeToggleForwarding(); }
    public boolean isConnected(int profileIndex)   { return NativeBridge.nativeIsConnected(profileIndex); }
    public int getProfileCount()                   { return NativeBridge.nativeGetProfileCount(); }
    public String getProfileName(int profileIndex) { return NativeBridge.nativeGetProfileName(profileIndex); }
    public int getActiveProfile()                  { return NativeBridge.nativeGetActiveProfile(); }

    public void setTtsEngine(String enginePackage) {
        ttsManager.shutdown();
        ttsManager = new TtsManager(getApplicationContext(), enginePackage);
        ttsManager.setScreenReaderMode(AppPrefs.getScreenReaderMode(this));
        ttsManager.setUseAccessibilityStream(AppPrefs.getAccessibilityStream(this));
        ttsManager.setPitch(AppPrefs.getTtsPitch(this));
        ttsManager.setRate(AppPrefs.getTtsRate(this));
        ttsManager.setVolume(AppPrefs.getTtsVolume(this));
        NativeBridge.nativeUpdateTtsManager(ttsManager);
        AppPrefs.setTtsEngine(this, enginePackage);
    }

    public void setUseAccessibilityStream(boolean use) {
        ttsManager.setUseAccessibilityStream(use);
        AppPrefs.setAccessibilityStream(this, use);
    }

    public void setScreenReaderMode(boolean use) {
        ttsManager.setScreenReaderMode(use);
        AppPrefs.setScreenReaderMode(this, use);
    }

    public void setTtsPitch(float value) {
        ttsManager.setPitch(value);
        AppPrefs.setTtsPitch(this, value);
    }

    public void setTtsRate(float value) {
        ttsManager.setRate(value);
        AppPrefs.setTtsRate(this, value);
    }

    public void setTtsVolume(float value) {
        ttsManager.setVolume(value);
        AppPrefs.setTtsVolume(this, value);
    }

    public void deleteProfile(int profileIndex) {
        desiredConnected.remove(profileIndex);
        executor.submit(() -> {
            if (NativeBridge.nativeIsConnected(profileIndex)) NativeBridge.nativeDisconnect(profileIndex);
            try {
                ProfileRepository.deleteProfile(profileIndex);
            } catch (Exception e) {
                Log.e(TAG, "Failed to delete profile " + profileIndex + ": " + e.getMessage());
            }
        });
    }

    private void autoConnect() {
        if (!AppPrefs.getAutoConnect(this)) return;
        executor.submit(() -> {
            int count = NativeBridge.nativeGetProfileCount();
            for (int i = 0; i < count; i++) {
                if (NativeBridge.nativeGetAutoConnect(i)) {
                    Log.i(TAG, "Auto-connecting profile " + i);
                    desiredConnected.add(i);
                    NativeBridge.nativeConnect(i);
                }
            }
        });
    }

    private void createNotificationChannel() {
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                getString(R.string.notification_channel_name),
                NotificationManager.IMPORTANCE_LOW);
        channel.setDescription(getString(R.string.notification_channel_desc));
        getSystemService(NotificationManager.class).createNotificationChannel(channel);
    }

    private Notification buildNotification(int connectedCount) {
        String text = connectedCount == 0
                ? getString(R.string.notification_no_connections)
                : getResources().getQuantityString(
                        R.plurals.notification_connected, connectedCount, connectedCount);

        PendingIntent openIntent = PendingIntent.getActivity(this, 0,
                new Intent(this, MainActivity.class),
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);

        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle(getString(R.string.app_name))
                .setContentText(text)
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentIntent(openIntent)
                .setOngoing(true)
                .setSilent(true)
                .build();
    }

    private void startForeground() {
        Notification n = buildNotification(0);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIFICATION_ID, n, ServiceInfo.FOREGROUND_SERVICE_TYPE_REMOTE_MESSAGING);
        } else {
            startForeground(NOTIFICATION_ID, n);
        }
    }

    private void updateNotification(Map<Integer, Boolean> states) {
        int count = 0;
        for (boolean v : states.values()) if (v) count++;
        getSystemService(NotificationManager.class).notify(NOTIFICATION_ID, buildNotification(count));
        if (count > 0) {
            if (!wakeLock.isHeld()) wakeLock.acquire();
        } else {
            if (wakeLock.isHeld()) wakeLock.release();
        }
    }
}

File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/service/ConnectionService.kt
package org.gozaltech.nvdaremotecompanion.android.service

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import android.util.Log
import androidx.core.app.NotificationCompat
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import org.gozaltech.nvdaremotecompanion.android.MainActivity
import org.gozaltech.nvdaremotecompanion.android.R
import org.gozaltech.nvdaremotecompanion.android.data.SettingsRepository
import org.gozaltech.nvdaremotecompanion.android.remote.NvdaRemoteRepository
import org.gozaltech.nvdaremotecompanion.android.speech.SpeechBridge
import org.koin.android.ext.android.inject

private const val TAG = "NVDARemote/Service"
private const val NOTIFICATION_ID = 1
private const val CHANNEL_ID = "nvdaremote_connection"
private const val WAKE_LOCK_TAG = "NVDARemote:connection"

private const val MEDIA_CAPABLE_SERVICE_TYPES =
    ServiceInfo.FOREGROUND_SERVICE_TYPE_REMOTE_MESSAGING or
        ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK

class ConnectionService : Service() {

    private val repository: NvdaRemoteRepository by inject()
    private val settingsRepository: SettingsRepository by inject()
    private val speech: SpeechBridge by inject()

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    private lateinit var wakeLock: PowerManager.WakeLock

    private val connectivityManager: ConnectivityManager
        get() = getSystemService(ConnectivityManager::class.java)

    private val notificationManager: NotificationManager
        get() = getSystemService(NotificationManager::class.java)

    private val networkCallback =
        object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = repository.reconnectDesired()

            override fun onLost(network: Network) = repository.suspendConnections()
        }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        wakeLock =
            getSystemService(PowerManager::class.java)
                .newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, WAKE_LOCK_TAG)

        val settings = settingsRepository.currentBlocking()
        speech.configure(settings)
        repository.start()
        if (settings.autoConnect) repository.autoConnectAll()

        connectivityManager.registerNetworkCallback(
            NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .build(),
            networkCallback,
        )

        scope.launch { settingsRepository.settings.collect(speech::configure) }
        scope.launch {
            repository.state
                .map { it.connectedCount }
                .distinctUntilChanged()
                .collect(::onConnectedCountChanged)
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val mediaCapable = intent?.getBooleanExtra(EXTRA_FOREGROUND_START, false) == true
        runCatching { startForegroundNotification(mediaCapable) }
            .onFailure { Log.w(TAG, "Could not enter the foreground: ${it.message}") }
        return START_STICKY
    }

    override fun onDestroy() {
        scope.cancel()
        runCatching { connectivityManager.unregisterNetworkCallback(networkCallback) }
        repository.shutdown()
        speech.shutdown()
        if (wakeLock.isHeld) wakeLock.release()
        super.onDestroy()
    }

    @SuppressLint("WakelockTimeout")
    private fun onConnectedCountChanged(connectedCount: Int) {
        notificationManager.notify(NOTIFICATION_ID, buildNotification(connectedCount))
        when {
            connectedCount > 0 && !wakeLock.isHeld -> wakeLock.acquire()
            connectedCount == 0 && wakeLock.isHeld -> wakeLock.release()
        }
    }

    private fun createNotificationChannel() {
        val channel =
            NotificationChannel(
                    CHANNEL_ID,
                    getString(R.string.notification_channel_name),
                    NotificationManager.IMPORTANCE_LOW,
                )
                .apply { description = getString(R.string.notification_channel_desc) }
        notificationManager.createNotificationChannel(channel)
    }

    private fun buildNotification(connectedCount: Int): Notification {
        val text =
            if (connectedCount == 0) {
                getString(R.string.notification_no_connections)
            } else {
                resources.getQuantityString(
                    R.plurals.notification_connected,
                    connectedCount,
                    connectedCount,
                )
            }
        val contentIntent =
            PendingIntent.getActivity(
                this,
                0,
                Intent(this, MainActivity::class.java),
                PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
            )
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(getString(R.string.app_name))
            .setContentText(text)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentIntent(contentIntent)
            .setOngoing(true)
            .setSilent(true)
            .build()
    }

    private fun startForegroundNotification(mediaCapable: Boolean) {
        val notification = buildNotification(repository.state.value.connectedCount)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            val types =
                if (mediaCapable) {
                    MEDIA_CAPABLE_SERVICE_TYPES
                } else {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_REMOTE_MESSAGING
                }
            startForeground(NOTIFICATION_ID, notification, types)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }
    }

    companion object {
        const val EXTRA_FOREGROUND_START = "foreground_start"

        fun startedFromUi(context: Context): Intent =
            Intent(context, ConnectionService::class.java).putExtra(EXTRA_FOREGROUND_START, true)

        fun startedFromBackground(context: Context): Intent =
            Intent(context, ConnectionService::class.java)
    }
}
